### PR TITLE
dev-python/pyzmq: Version bump to 17.1.0

### PR DESCRIPTION
**Pull request by marbre**
Unfortunately,  the test `lib/zmq/tests/test_pubsub.py::TestPubSubGreen::test_sigabrt_issue` hangs and I have no idea why.

Closes: https://bugs.gentoo.org/648088
Package-Manager: Portage-2.3.40, Repoman-2.3.9

### CODE REVIEW

**Comment by sbraz** [comment]
Or exclude the failing test

**Comment by sbraz** [comment]
> Unfortunately, the test lib/zmq/tests/test_pubsub.py::TestPubSubGreen::test_sigabrt_issue hangs and I have no idea why.

Those were apparently removed in https://github.com/zeromq/pyzmq/commit/e9000fc82ca799f6de73573b51c6b73fb519b6f3 which is included in v17.1.0 (it probably wasn't released when you made the previous PR). Can you bump to 17.1.0?

**Comment by marbre** [comment]
@sbraz Yes, 17.0.1 originally wasn't released. I'll bump to 17.0.1 now. Anyway, now some other tests fail:

```
========================================================================= FAILURES ==========================================================================
_____________________________________________________________ TestZMQStream.test_on_recv_basic ______________________________________________________________

self = <zmq.tests.test_zmqstream.TestZMQStream testMethod=test_on_recv_basic>

    def setUp(self):
        if tornado is None:
            pytest.skip()
        if asyncio:
            asyncio.set_event_loop(asyncio.new_event_loop())
        self.context = zmq.Context()
        self.loop = ioloop.IOLoop.instance()
>       self.push = zmqstream.ZMQStream(self.context.socket(zmq.PUSH))

../pyzmq-17.1.0-python3_4/lib/zmq/tests/test_zmqstream.py:31: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
../pyzmq-17.1.0-python3_4/lib/zmq/eventloop/zmqstream.py:120: in __init__
    self._init_io_state()
../pyzmq-17.1.0-python3_4/lib/zmq/eventloop/zmqstream.py:541: in _init_io_state
    self.io_loop.add_handler(self.socket, self._handle_events, self.io_loop.READ)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <zmq.eventloop.ioloop.ZMQIOLoop object at 0x7f64b0055940>, fd = 150
handler = <bound method ZMQStream._handle_events of <zmq.eventloop.zmqstream.ZMQStream object at 0x7f64b0052630>>, events = 1

    def add_handler(self, fd, handler, events):
        fd, obj = self.split_fd(fd)
        self._handlers[fd] = (obj, stack_context.wrap(handler))
>       self._impl.register(fd, events | self.ERROR)
E       ValueError: I/O operation on closed epoll object

/usr/lib64/python3.4/site-packages/tornado/ioloop.py:728: ValueError
______________________________________________________________ TestZMQStream.test_on_recv_wake ______________________________________________________________

self = <zmq.tests.test_zmqstream.TestZMQStream testMethod=test_on_recv_wake>

    def setUp(self):
        if tornado is None:
            pytest.skip()
        if asyncio:
            asyncio.set_event_loop(asyncio.new_event_loop())
        self.context = zmq.Context()
        self.loop = ioloop.IOLoop.instance()
>       self.push = zmqstream.ZMQStream(self.context.socket(zmq.PUSH))

../pyzmq-17.1.0-python3_4/lib/zmq/tests/test_zmqstream.py:31: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
../pyzmq-17.1.0-python3_4/lib/zmq/eventloop/zmqstream.py:120: in __init__
    self._init_io_state()
../pyzmq-17.1.0-python3_4/lib/zmq/eventloop/zmqstream.py:541: in _init_io_state
    self.io_loop.add_handler(self.socket, self._handle_events, self.io_loop.READ)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <zmq.eventloop.ioloop.ZMQIOLoop object at 0x7f64b0055940>, fd = 159
handler = <bound method ZMQStream._handle_events of <zmq.eventloop.zmqstream.ZMQStream object at 0x7f64b00c1f28>>, events = 1

    def add_handler(self, fd, handler, events):
        fd, obj = self.split_fd(fd)
        self._handlers[fd] = (obj, stack_context.wrap(handler))
>       self._impl.register(fd, events | self.ERROR)
E       ValueError: I/O operation on closed epoll object

/usr/lib64/python3.4/site-packages/tornado/ioloop.py:728: ValueError
===================================================== 2 failed, 314 passed, 10 skipped in 67.87 seconds =====================================================#
```

As `tornado` is involved, it might be that the stable version in portage is just too old. I'll skip those two tests. Further, with `pyhon3.6` the `test_buffer_numpy` hangs, which I will therefore disabled for 3.6.

**Comment by mgorny** [comment]
Link upstream bug reports for those failures, please.

**Comment by marbre** [comment]
@mgorny Linked upstream bug reports within the patches.

**Comment by marbre** [comment]
Dropped one of the patches and raised the required tornado version.

**Comment by marbre** [comment]
@sbraz @mgorny Any further changes required or is this ready to merge?

**Comment by marbre** [comment]
@mgorny No problem, squashed them :)

**Comment by mgorny** [comment]
Just found another bug:

```cc /var/tmp/portage/dev-python/pyzmq-17.1.0/work/pyzmq-17.1.0-python2_7/temp.linux-x86_64-2.7/scratch/var/tmp/portage/dev-python/pyzmq-17.1.0/temp/timer_createndjAA4.o -o /var/tmp/portage/dev-python/pyzmq-17.1.0/work/pyzmq-17.1.0-python2_7/temp.linux-x86_64-2.7/scratch/a.out
/var/tmp/portage/dev-python/pyzmq-17.1.0/work/pyzmq-17.1.0-python2_7/temp.linux-x86_64-2.7/scratch/var/tmp/portage/dev-python/pyzmq-17.1.0/temp/timer_createndjAA4.o: In function `main':
timer_createndjAA4.c:(.text+0x15): undefined reference to `timer_create'
collect2: error: ld returned 1 exit status
```

Note that it calls `cc` instead of proper `$(CC)`. If you could investigate if that's problem with this package's build system or sth else, I'd appreciate that. Since it's not a regression (or at least I doubt it is), I'm going to merge it nevertheless once I finish testing.

**Comment by mgorny** [comment]
It seems that `../pyzmq-17.1.0-python3_6/lib/zmq/tests/test_message.py::TestFrame::test_buffer_numpy` hanged for me this time. I can retry tomorrow, if you can't reproduce it.

**Comment by marbre** [comment]
`test_buffer_numpy` hangs heree to with python3.6 and is therefore excluded. Which python version did you use?

**Comment by mgorny** [comment]
What do you mean 'excluded'? 3.6.6 is installed.

**Comment by marbre** [comment]
Regarding `timer_create`, it is checked within `setup.py` if cc provides `cc.has_function('timer_create'):`. It not `libtrt` should be linked. Have to dig in deeper why this isn't working/called. The problem also persists with `pyzmq-16.0.2`.

**Comment by marbre** [comment]
The patch I added should skip this test for python3.6.

**Comment by mgorny** [comment]
> Regarding `timer_create`, it is checked within `setup.py` if cc provides `cc.has_function('timer_create'):`. It not `libtrt` should be linked. Have to dig in deeper why this isn't working/called. The problem also persists with `pyzmq-16.0.2`.

Well, I think it is called but it uses `cc` instead of `x86_64-blahblahblah-gcc` ;-).

File: dev-python/pyzmq/files/pyzmq-17.1.0-test_message.patch [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+ 
++import pytest
+ import zmq
+ from zmq.tests import BaseZMQTestCase, SkipTest, skip_pypy, PYPY
+ from zmq.utils.strtypes import unicode, bytes, b, u
+@@ -262,6 +263,7 @@ class TestFrame(BaseZMQTestCase):
+             self.assertEqual(s,r)
+         self.assertEqual(s, m.bytes)
+     
++    @pytest.mark.skipif(sys.version_info == (3,6))
```
Comment (mgorny) [inline dev-python/pyzmq/files/pyzmq-17.1.0-test_message.patch]: FYI:

```
>>> import sys
>>> sys.version_info
sys.version_info(major=3, minor=6, micro=6, releaselevel='final', serial=0)
>>> sys.version_info == (3,6)
False
>>> sys.version_info == (3,6,6)
False
>>> sys.version_info == (3,6,6,'final',0)
True

**Comment by marbre** [comment]
Any hint in which way we would than have to patch `cc = new_compiler(compiler=self.compiler_type)`?

File: dev-python/pyzmq/files/pyzmq-17.1.0-test_message.patch (hunk 1 above)
Comment (marbre) [reply dev-python/pyzmq/files/pyzmq-17.1.0-test_message.patch]: Thanks. Will rework the patch in some way.

**Comment by mgorny** [comment]
Apparently distutils is doing:

```
customize_compiler(self.compiler)  # from sysconfig
```

immediately after creating `new_compiler()`.

**Comment by mgorny** [comment]
Do you mean that the customize_compiler thing did not work?

**Comment by marbre** [comment]
At least I did not managed it to get working as expected. Seems that I didn't checked the setup.py well enough. The checking for `librt` is done within `bundle_libzmq_extension`, called by `fallback_on_bundled`. However, the bundled zmq is only used if zmq is not found on the system. Thus, I don't think this is the problem at all.

Checking the error message again
```
/var/tmp/portage/dev-python/pyzmq-17.1.0/temp/timer_createh3iwlpck.c: In function ‘main’:
/var/tmp/portage/dev-python/pyzmq-17.1.0/temp/timer_createh3iwlpck.c:2:5: warning: implicit declaration of function ‘timer_create’ [-Wimplicit-function-declaration]
     timer_create();
     ^~~~~~~~~~~~
cc /var/tmp/portage/dev-python/pyzmq-17.1.0/work/pyzmq-17.1.0-python3_4/temp.linux-x86_64-3.4/scratch/var/tmp/portage/dev-python/pyzmq-17.1.0/temp/timer_createh3iwlpck.o -o /var/tmp/portage/dev-python/pyzmq-17.1.0/work/pyzmq-17.1.0-python3_4/temp.linux-x86_64-3.4/scratch/a.out
/var/tmp/portage/dev-python/pyzmq-17.1.0/work/pyzmq-17.1.0-python3_4/temp.linux-x86_64-3.4/scratch/var/tmp/portage/dev-python/pyzmq-17.1.0/temp/timer_createh3iwlpck.o: In function `main':
timer_createh3iwlpck.c:(.text+0x15): undefined reference to `timer_create'
collect2: error: ld returned 1 exit status
```
I took a look into `/var/tmp/portage/dev-python/pyzmq-17.1.0/temp/timer_createh3iwlpck.c`, only containing
```c
main (int argc, char **argv) {
    timer_create();
}
```
Thus, as stated in the error message above, it's an missing include. Couldn't find out where this file (not included in the source bundle) is generated and what's really going on, and probably won't have the time to do so.

**Comment by mgorny** [comment]
I've sent a pull request upstream for that CC thing. It's not a major issue, so no point in backporting the fix.

However, I'm afraid the entropy is not in your favor today. This time that numpy test hanged with py3.5 on me. Given that it's unpredictable, I think it'd be both better and simpler to just sed it out unconditionally (add `_` before the method name).

Would you be able to do that?

**Comment by marbre** [comment]
@mgorny Sure, replaced `skipif` by `skip`.

**Comment by marbre** [comment]
@mgorny @monsieurp Ready to merge in order to solve some of the related bugs which depend on this version bump?

**Comment by mgorny** [comment]
I'm sorry, the endless chromium rebuilds are delaying me from testing this. The fact that some of the tests eat all my memory isn't helping.

**Comment by mgorny** [comment]
Ok, I'm starting another test run and closing everything else to prevent it from OOM-ing me hard. If it succeeds, I'll push it. If it doesn't, I'll let you know tomorrow.

**Comment by mgorny** [comment]
…aaaaand merge!

That said, the package recompiles all extensions in `src_install` which is bad. Curious enough, the code compiled for py2.7 is *different* than the one used in `src_compile`, judging by ccache not working there — I'd guess it's been cythonized for Python 3. I think there was a distutils option to disable recompiling in `install` command, you probably want to use that.

**Pull request closed without merging**